torch/csrc/jit/codegen/cuda/fusion_segmenter.h

```diff
@@ -0,0 +1,365 @@
+// already a DAG). The segmentation code relies on assumptions of DAG-ness
+// during segmentation, meaning proposed merging of groups must maintain the DAG
+// property of the graph.
+//
+// Julien Herrmann, Yusuf Özkaya, Bora Uçar, Kamer Kaya, Umit Catalyurek.
+// Multilevel Algorithms for Acyclic Partitioning of Directed Acyclic Graphs.
+// SIAM Journal on Scientific Computing, Society for Industrial and Applied
+// Mathematics, 2019, 41 (4), pp.A2117-A2145. ff10.1137/18M1176865ff.
+// ffhal02306566f
+
+class SegmentedGroup;
+class SegmentCandidateFinder;
+
+// A directed edge on DAG,
+// Wrapper for values, edges between segmented groups which are made up
+// of Exprs. Multiple edges can exist between segmented groups.
+class SegmentedEdge {
+ public:
+  SegmentedEdge(SegmentedGroup* from, SegmentedGroup* to, Val* val)
+      : from_(from), to_(to), val_(val) {}
+  SegmentedGroup* from_;
+  SegmentedGroup* to_;
+  Val* val_;
+
+  void print();
+};
+
+std::ostream& operator<<(std::ostream& os, const SegmentedEdge* edge);
+
+// Groups together expressions which create a segmented group
+// Can be used to produce fusions
+class TORCH_CUDA_API SegmentedGroup {
+ public:
+  SegmentedGroup() = default;
+
+  SegmentedGroup(Expr* expr) {
+    exprs_.push_back(expr);
+  }
+
+  void clearTraversalInfo();
+
+  std::vector<SegmentedGroup*> getNeighbors();
+
+  // TODO: May want to sort this based on size of connections between this and
+  // neighbors as well as if the connection is an output of the fusion (has to
+  // be saved to gmem anyways)
+  std::vector<std::pair<SegmentedGroup*, SegmentedEdge*>> getNeighborsPair();
+
+  // Look at all neighbors of this and return who this could merge with based on
+  // level values of this, neighbors, and merged neighbors of neighbors
+  std::vector<std::pair<SegmentedGroup*, SegmentedEdge*>> getMergeCandidates();
+
+  bool isInputGroup() {
+    return !input_vals.empty();
+  };
+
+  bool isConnected() const {
+    return !producer_edges.empty() || !consumer_edges.empty();
+  }
+
+  int groupId() const {
+    return group_id_;
+  }
+
+  // Consider Passkey or friend for safety
+  void setID(int id) {
+    TORCH_INTERNAL_ASSERT(group_id_ == -1);
+    group_id_ = id;
+  }
+
+  std::vector<Val*> inputs() {
+    return input_vals;
+  }
+
+  std::vector<Val*> outputs() {
+    return output_vals;
+  }
+
+  // might just set heuristic as public
+  ScheduleHeuristic heuristic() {
+    return heuristic_;
+  }
+
+  void setHeuristic(ScheduleHeuristic sh) {
+    heuristic_ = sh;
+  }
+
+  void print();
+
+  //! Generate the use info for groups to avoid multi-use in
+  //!  a single group, expect to remove in the future so
+  //!  keeping this part separate
+  void updateUse();
+
+  //! Check if an edge is used internally within a group
+  bool hasUse(Val* se);
+
+  //! To be called at the very end of segment fusion
+  //!  no more segment merging should be done beyond
+  void finalize();
+
+ public:
```

Comment:
this section seems to be a mix of public and seemingly private data - can we clean this up?

(ex. make all private and use the _ suffix consistently)

Reply:
Did a clean up pass on this class and hope it looks better now. Thanks!